widgets/src/input/text_selectable.rs

```diff
@@ -0,0 +1,92 @@
+use super::CaretState;
+use super::{glyphs_helper::GlyphsHelper, selected_text::SelectedText};
+use crate::layout::{Stack, StackFit};
+use crate::prelude::Text;
+use ribir_core::prelude::*;
+
+#[derive(Declare, Default)]
+pub struct TextSelectable {
+  #[declare(default)]
+  pub caret: CaretState,
+  #[declare(skip, default)]
+  pub(crate) helper: GlyphsHelper,
+}
+
+impl ComposeChild for TextSelectable {
+  type Child = State<Text>;
+  fn compose_child(this: State<Self>, text: Self::Child) -> Widget {
+    let this = this.into_writable();
+    widget! {
+      states {
+        this: this.clone(),
+        text: text.into_readonly(),
+      }
+      Stack {
+        fit: StackFit::Passthrough,
+        on_pointer_move: move |e| {
+          if let CaretState::Selecting(begin, _) = this.caret {
+            if e.point_type == PointerType::Mouse
+              && e.mouse_buttons() == MouseButtons::PRIMARY {
+              let position = e.position();
+              let cluster = this.helper.cluster_from_pos(position.x, position.y);
+              this.caret = CaretState::Selecting(begin, cluster as usize);
+            }
+          }
+        },
+        on_pointer_down: move |e| {
+          let position = e.position();
+          let cluster = this.helper.cluster_from_pos(position.x, position.y);
+          this.caret = CaretState::Selecting(cluster as usize, cluster as usize);
+        },
+        on_pointer_up: move |_| {
+          if let CaretState::Selecting(begin, end) = this.caret {
+            this.caret = if begin == end {
+              CaretState::Caret(begin)
+            } else {
+              CaretState::Select(begin, end)
+            };
+          }
+        },
+
+        on_key_down: move |event| {
+            if event.key == VirtualKeyCode::C && event.common.with_command_key(){
```

Comment:
```suggestion
            if event.key == VirtualKeyCode::C && event.common.with_command_key(){
```
code indentation 2 spaces is better